### PR TITLE
Fix comparing bundle versions with ocp versions [CLOUDDST-10476]

### PR DESCRIPTION
**Pull request by pbortlov**
- modify version comparison
- add packaging library
- add debug comments

### CODE REVIEW

File: iib/workers/tasks/build_merge_index_image.py [hunk 1]
```diff
@@ -200,8 +200,8 @@ def handle_merge_request(
             raise IIBError(err_msg)
 
         set_request_state(request_id, 'in_progress', 'Getting bundles present in the index images')
-        log.info('Getting bundles present in the source index image')
         with set_registry_token(overwrite_target_index_token, source_from_index):
+            log.info('Getting bundles present in the source index image')
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/build_merge_index_image.py:204]: Is this change required?

File: iib/workers/tasks/build_merge_index_image.py [hunk 2]
```diff
@@ -344,3 +345,13 @@ def is_bundle_version_valid(bundle_path, valid_ocp_version):
         )
 
     return False
+
+def correct_float_precision(version):
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/build_merge_index_image.py:349]: Maybe rename it to simply `float_precision` and have the precision as a parameter which defaults to 2?

**Review by yashvardhannanavati — CHANGES_REQUESTED** [review]

File: iib/workers/tasks/build_merge_index_image.py [hunk 3]
```diff
@@ -317,24 +317,25 @@ def is_bundle_version_valid(bundle_path, valid_ocp_version):
...
+            min_version, max_version = [correct_float_precision(version) for version in bundle_version.split('-')]
             if min_version <= float_valid_ocp_version <= max_version:
                 return True
         elif "," in bundle_version_label:
-            versions = [float(version) for version in bundle_version.split(",")]
+            versions = [correct_float_precision(version) for version in bundle_version.split(",")]
             if float_valid_ocp_version >= versions[0]:
                 return True
-        elif float_valid_ocp_version >= float(bundle_version):
+        elif float_valid_ocp_version >= correct_float_precision(bundle_version):
```
Comment (lcarva) [inline iib/workers/tasks/build_merge_index_image.py:338]: I don't think this will work as expected:

```python
>>> '4.10' > '4.9'
False
```

File: iib/workers/tasks/build_merge_index_image.py [hunk 4]
```diff
@@ -317,24 +317,25 @@ def is_bundle_version_valid(bundle_path, valid_ocp_version):
     v4.7   | included  | NOT included |  included   |  included    |  included
     """
```
Comment (lcarva) [inline iib/workers/tasks/build_merge_index_image.py:318]: Let's update the unit tests

File: iib/workers/tasks/build_merge_index_image.py (hunk 2 above)
Comment (pbortlov) [reply iib/workers/tasks/build_merge_index_image.py:349]: sure

File: iib/workers/tasks/build_merge_index_image.py (hunk 3 above)
Comment (lcarva) [reply iib/workers/tasks/build_merge_index_image.py:338]: I think you'll need something like this: https://gist.github.com/limed/473a498641bbc7761a20

Comment (pbortlov) [reply iib/workers/tasks/build_merge_index_image.py]: I see. We can't compare those numbers

File: iib/workers/tasks/build_merge_index_image.py [hunk 5]
```diff
@@ -344,3 +357,15 @@ def is_bundle_version_valid(bundle_path, valid_ocp_version):
...
+def sort_floats(versions):
+    """
+    Sort list of strings with floats
+
+    :param list versions: unsorted list of versions
+    :rtype: list
+    """
+    convert = lambda text: int(text) if text.isdigit() else text
+    alphanum_key = lambda key: [convert(c) for c in re.split('([0-9]+)', key)]
+    versions.sort(key=alphanum_key)
```
Comment (lcarva) [inline iib/workers/tasks/build_merge_index_image.py:363]: This sorts the given list in place. Either use `sorted` instead, or don't return the list.

File: iib/workers/tasks/build_merge_index_image.py [hunk 6]
```diff
@@ -308,33 +309,38 @@ def is_bundle_version_valid(bundle_path, valid_ocp_version):
...
         if bundle_version_label.startswith('='):
-            if float(bundle_version.strip('=')) == float_valid_ocp_version:
+            if bundle_version.strip('=') == ocp_version:
                 return True
         elif '-' in bundle_version_label:
-            min_version, max_version = [float(version) for version in bundle_version.split('-')]
-            if min_version <= float_valid_ocp_version <= max_version:
+            min_version, max_version = [version for version in bundle_version.split('-')]
+
+            if sort_floats(list(min_version + ocp_version)) == list(min_version + ocp_version) and \
```
Comment (lcarva) [inline iib/workers/tasks/build_merge_index_image.py:336]: The usage of lists seems unnatural here. Consider this. Instead of using `sort_floats`, introduce a new function, e.g. `to_version_tuple`, which does the conversion that is currently embedded in `sort_floats`, e.g. `'v4.10'` -> `(4, 10)`. Then, you can just compare the tuples as you'd compare any other value. `sort_floats` wouldn't be needed.

File: iib/workers/tasks/build_merge_index_image.py (hunk 1 above)
Comment (yashvardhannanavati) [reply iib/workers/tasks/build_merge_index_image.py:205]: Probably should be a separate commit. fwiw,  I think it wouldn't matter if we bring it 2 lines below or keep it as is :)

**Review by yashvardhannanavati — CHANGES_REQUESTED** [review]

File: iib/workers/tasks/build_merge_index_image.py (hunk 6 above)
Comment (pbortlov) [reply iib/workers/tasks/build_merge_index_image.py:336]: what do you think about this? https://codereview.stackexchange.com/questions/217992/compare-two-version-numbers-in-the-form-major-minor-patch-build-to-see-if-one-is

File: iib/workers/tasks/build_merge_index_image.py (hunk 1 above)
Comment (pbortlov) [reply iib/workers/tasks/build_merge_index_image.py]: removed

File: iib/workers/tasks/build_merge_index_image.py (hunk 6 above)
Comment (lcarva) [reply iib/workers/tasks/build_merge_index_image.py]: Ooh I like that! `LooseVersion` should do the trick for us.

```python
>>> from distutils.version import LooseVersion
>>> LooseVersion('4.10') > LooseVersion('4.9')
True
>>> LooseVersion('4.1') > LooseVersion('4.9')
False
>>> LooseVersion('4.10') > LooseVersion('4.101')
False
>>> LooseVersion('4.10') > LooseVersion('4.09')
True
```

Comment (pbortlov) [reply iib/workers/tasks/build_merge_index_image.py]: IT seems that this already deprecated. We can use this:
>>> from packaging import version
>>> version.parse("2.3.1") < version.parse("10.1.2")
True
>>> version.parse("1.3.a4") < version.parse("10.1.2")
True
>>> isinstance(version.parse("1.3.a4"), version.Version)
True
>>> isinstance(version.parse("1.3.xy123"), version.LegacyVersion)
True
>>> version.Version("1.3.xy123")
Traceback (most recent call last):
...
packaging.version.InvalidVersion: Invalid version: '1.3.xy123'

File: iib/workers/tasks/build_merge_index_image.py (hunk 4 above)
Comment (pbortlov) [reply iib/workers/tasks/build_merge_index_image.py:322]: Done

**Review by lcarva — DISMISSED** [review]

**Review by yashvardhannanavati — APPROVED** [review]

File: iib/workers/tasks/build_merge_index_image.py [hunk 7]
```diff
@@ -308,33 +310,36 @@ def is_bundle_version_valid(bundle_path, valid_ocp_version):
...
-            min_version, max_version = [float(version) for version in bundle_version.split('-')]
-            if min_version <= float_valid_ocp_version <= max_version:
+            min_version, max_version = [Version(version) for version in bundle_version.split('-')]
+            if min_version <= ocp_version <= max_version:
                 return True
         elif "," in bundle_version_label:
-            versions = [float(version) for version in bundle_version.split(",")]
-            if float_valid_ocp_version >= versions[0]:
+            versions = [Version(version) for version in bundle_version.split(",")]
+            if versions[0] <= ocp_version:
```
Comment (chandwanitulsi) [inline iib/workers/tasks/build_merge_index_image.py:340]: The comparisons are reversed here, but that shouldn't a problem

**Review by chandwanitulsi — APPROVED** [review]
